flink/src/main/java/org/apache/iceberg/flink/sink/FlinkManifest.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.ManifestFile;
+import org.apache.iceberg.ManifestFiles;
+import org.apache.iceberg.ManifestWriter;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+class FlinkManifest {
+  private static final int ICEBERG_FORMAT_VERSION = 2;
+  private static final Long DUMMY_SNAPSHOT_ID = 0L;
+
+  private final OutputFile outputFile;
+  private final PartitionSpec spec;
+
+  private FlinkManifest(OutputFile outputFile, PartitionSpec spec) {
+    this.outputFile = outputFile;
+    this.spec = spec;
+  }
+
+  ManifestFile write(List<DataFile> dataFiles) throws IOException {
+    ManifestWriter<DataFile> writer = ManifestFiles.write(ICEBERG_FORMAT_VERSION, spec, outputFile, DUMMY_SNAPSHOT_ID);
+    try (ManifestWriter<DataFile> closeableWriter = writer) {
+      closeableWriter.addAll(dataFiles);
+    }
+    return writer.toManifestFile();
+  }
+
+  static List<DataFile> read(ManifestFile manifestFile, FileIO io) throws IOException {
+    try (CloseableIterable<DataFile> dataFiles = ManifestFiles.read(manifestFile, io).project(ManifestFile.schema())) {
+      return Lists.newArrayList(dataFiles);
+    }
+  }
+
+  static FlinkManifestFactory createFactory(Table table, String flinkJobId, int partitionId, long taskId) {
+    return new FlinkManifestFactory(table.location(), table.spec(), FileFormat.AVRO, table.locationProvider(),
+        table.io(), table.encryption(), flinkJobId, partitionId, taskId);
+  }
+
+  static class FlinkManifestFactory {
+    private final String tableLocation;
+    private final PartitionSpec spec;
+    private final FileFormat format;
+    private final OutputFileFactory outputFileFactory;
+    private final int partitionId;
+    private final long taskId;
+    private final String flinkJobId;
+    private final AtomicInteger fileCount = new AtomicInteger(0);
+
+    FlinkManifestFactory(String tableLocation, PartitionSpec spec, FileFormat format, LocationProvider locations,
+                         FileIO io, EncryptionManager encryptionManager, String flinkJobId, int partitionId,
+                         long taskId) {
+      this.tableLocation = tableLocation;
+      this.spec = spec;
+      this.format = format;
+      this.flinkJobId = flinkJobId;
+      this.partitionId = partitionId;
+      this.taskId = taskId;
+      this.outputFileFactory = new OutputFileFactory(spec, format, locations, io,
+          encryptionManager, partitionId, taskId);
+    }
+
+    private String generateRelativeFilePath(long checkpointId) {
```

Comment:
just a note for future improvement, we probably need to allow users to supply the manifest file base path. E.g. we run the streaming Flink jobs in 3 different regions, while Iceberg table's base path is only in one region. We don't want to write the manifiest file to the table home/base path. Instead, we want to write it to a local region (S3 bucket). I haven't thought about too much on how we can extend from the open source implementation yet. We probably can discuss it in a more systematic/thorough way maybe when we integrating Iceberg sink with the new FLIP-143 sink interface.

Reply:
+1 to this.